core/nandtool/nandtool.c

```diff
@@ -0,0 +1,245 @@
+/*
+ * Phoenix-RTOS
+ *
+ * NANDtool utility
+ *
+ * Allows to program/erase NAND flash memory
+ *
+ * Copyright 2018, 2021 Phoenix Systems
+ * Author: Kamil Amanowicz, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/file.h>
+#include <sys/msg.h>
+#include <sys/stat.h>
+
+#include "flashmng.h"
+
+
+static struct {
+	oid_t oid;
+	int fd;
+	flashsrv_info_t *info;
+} nandtool_common;
+
+
+static int nandtool_flash(const char *path, unsigned int start, int raw)
+{
+	const flashsrv_info_t *info = nandtool_common.info;
+	const unsigned int npages = info->erasesz / info->writesz;
+	const unsigned int pagesz = raw ? info->writesz + info->metasz : info->writesz;
+	unsigned int page, block, boffs, offs = 0;
+	struct stat stat;
+	char *buff;
+	int fd, err;
+
+	if ((err = open(path, O_RDONLY)) < 0) {
+		fprintf(stderr, "nandtool: failed to open file %s, err: %d\n", path, err);
+		return err;
+	}
+	fd = err;
+
+	if ((buff = malloc(pagesz)) == NULL) {
+		err = -ENOMEM;
+		fprintf(stderr, "nandtool: failed to allocate buffer, err: %d\n", err);
+		close(fd);
+		return err;
+	}
+
+	do {
+		if ((err = fstat(fd, &stat)) < 0) {
+			fprintf(stderr, "nandtool: failed to stat file %s, err: %d\n", path, err);
+			break;
+		}
+
+		start *= npages;
+		while (offs < stat.st_size) {
+			page = start + (offs / pagesz);
+			block = page / npages;
+
+			if ((err = flashmng_isbad(nandtool_common.oid, block)) < 0) {
+				fprintf(stderr, "nandtool: failed to check block %u, err: %d\n", block, err);
+				break;
+			}
+			else if (err > 0) {
+				start++;
+				continue;
+			}
+
+			memset(buff, 0, pagesz);
+			for (boffs = 0; boffs < pagesz; boffs += err) {
+				if ((err = read(fd, buff + boffs, pagesz - boffs)) <= 0) {
+					if (err) {
+						err = -errno;
+						fprintf(stderr, "nandtool: failed to read file %s, err: %d\n", path, err);
+					}
+					break;
+				}
+			}
+
+			if (err < 0)
+				break;
+
+			if (raw) {
+				if ((err = flashmng_writeraw(nandtool_common.oid, page, buff, pagesz)) < 0) {
+					fprintf(stderr, "nandtool: failed to write raw data to page %u, err: %d\n", page, err);
+					break;
+				}
+			}
+			else {
+				if ((err = lseek(nandtool_common.fd, (off_t)page * pagesz, SEEK_SET)) < 0) {
```

Comment:
- err is of type int, but `lseek` returns `off_t`. If we seek further than 2GB, we will have negative value here
 - `lseek` returns error by `errno`

Reply:
Fixed.